terraform/modules/opensearch/data.tf

```diff
@@ -1,22 +1,94 @@
-data "aws_region" "region" {}
+data "aws_caller_identity" "current" {}
 
-data "aws_caller_identity" "caller" {}
+data "aws_iam_policy_document" "logs" {
+  count = var.create_cloudwatch_log_policy ? 1 : 0
 
-data "aws_iam_policy_document" "os" {
   statement {
     effect = "Allow"
     actions = [
       "logs:PutLogEvents",
-      "logs:PutLogEventsBatch",
       "logs:CreateLogStream"
     ]
     principals {
       type        = "Service"
       identifiers = ["es.amazonaws.com"]
     }
     resources = [
-      aws_cloudwatch_log_group.os.arn,
-      "${aws_cloudwatch_log_group.os.arn}:*"
+      "${aws_cloudwatch_log_group.this[0].arn}",
+      "${aws_cloudwatch_log_group.this[0].arn}:*"
+    ]
+  }
+}
+
+data "aws_iam_policy_document" "access_policy" {
+  count = var.create_access_policies ? 1 : 0
+
+  statement {
+    effect = "Allow"
+    actions = [
+      "es:ESHttpPut",
+      "es:ESHttpPost",
+      "es:ESHttpPatch",
+      "es:ESHttpHead",
+      "es:ESHttpGet",
+      "es:ESHttpDelete"
+    ]
+    principals {
+      type        = "AWS"
+      identifiers = ["*"]
+    }
+    resources = ["${aws_opensearch_domain.this.arn}/*"]
+  }
+}
+
+data "aws_iam_policy_document" "trust" {
+  count = var.create_snapshot_role ? 1 : 0
+
+  statement {
+    effect  = "Allow"
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["es.amazonaws.com"]
+    }
+  }
+}
+
+data "aws_iam_policy_document" "snapshot" {
+  count = var.create_snapshot_role ? 1 : 0
+
+  statement {
+    effect    = "Allow"
+    actions   = ["s3:ListBucket"]
+    resources = [var.s3_snapshot_bucket_arn]
+  }
+
+  statement {
+    effect = "Allow"
+    actions = [
+      "s3:GetObject",
+      "s3:PutObject",
+      "s3:DeleteObject"
+    ]
+    resources = ["${var.s3_snapshot_bucket_arn}/*"]
+  }
+
+  statement {
+    effect = "Allow"
+    actions = [
+      "iam:PassRole",
+      "iam:GetRole"
+    ]
+    resources = ["arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/power-user*"]
```

Comment:
should this reference the roles created in this module? I don't anticipate a change from the "power_user" naming convention, but if this were to happen this would be one less place this is hardcoded. alternatively we generally have a variable "iam_prefix" that is set by default to "power-user", this could be used here

Reply:
- Set the value in the resource block for this statement to reference the role conditionally created to support pass/get role actions during snapshot operations.